jax/_src/dtypes.py

```diff
@@ -69,52 +71,55 @@
     ]
 }
 
-def to_numeric_dtype(dtype):
+def to_numeric_dtype(dtype: DTypeLike) -> DType:
   """Promotes a dtype into an numeric dtype, if it is not already one."""
-  dtype = np.dtype(dtype)
-  return np.dtype('int32') if dtype == np.dtype('bool') else dtype
+  dtype_ = np.dtype(dtype)
+  return np.dtype('int32') if dtype_ == np.dtype('bool') else dtype_
 
 
-def to_inexact_dtype(dtype):
+def to_inexact_dtype(dtype: DTypeLike) -> DType:
   """Promotes a dtype into an inexact dtype, if it is not already one."""
-  dtype = np.dtype(dtype)
-  return _dtype_to_inexact.get(dtype, dtype)
+  dtype_ = np.dtype(dtype)
+  return _dtype_to_inexact.get(dtype_, dtype_)
 
 
-def to_complex_dtype(dtype):
+def to_complex_dtype(dtype: DTypeLike) -> DType:
   ftype = to_inexact_dtype(dtype)
   if ftype in [np.dtype('float64'), np.dtype('complex128')]:
     return np.dtype('complex128')
   return np.dtype('complex64')
 
 
 @functools.lru_cache(maxsize=None)
-def _canonicalize_dtype(x64_enabled, dtype):
+def _canonicalize_dtype(x64_enabled: bool, dtype: DTypeLike) -> DType:
   """Convert from a dtype to a canonical dtype based on config.x64_enabled."""
   if jax.core.is_opaque_dtype(dtype):
-    return dtype
+    # TODO(frostig): opaque dtypes break the return type guarantee here... what to do?
+    return dtype  # type: ignore[return-value]
```

Comment:
@froystig requesting your review in case you have thoughts about how to handle this.

Reply:
Opaque dtypes are unlike dtypes—both structurally and nominally. Should we annotate the function as returning `Union[DType, jax._src.prng.KeyTy, jax.core.bint]`?

cc @mattjj

Reply:
If it would be helpful to track the various opaque dtype definitions across the codebase, we could consider:
1. making them all conform to a protocol,
2. having them all derive from a common base type, or
3. manually maintaining an alias to a `Union` of them in `jax._src.typing`, e.g. named `OpaqueDType`

Reply:
The problem with returning a union here is that it would pollute basically every dtype in JAX – we use `canonicalize_dtype` in many places where the output is required to be `DType`.

How frequently are opaque dtypes used? Could we have a special version of `canonicalize_dtype` to use in the places where opaque dtypes are expected?

Reply:
Good idea to see whether we can distinguish at the call site. We'd have to investigate... The changes were introduced in #11768, which added only one explicit seemingly-relevant call to `canonicalize_dtype` (in changes to `api_util._shaped_abtractify_slow`). It's still possible that we rely on this change to `canonicalize_dtype` in other places, but this is where I'd look first to see whether modifying a single caller suffices.

Reply:
Thanks, I'll try adding an `allow_opaque_dtypes` keyword to `canonicalize_dtype` that defaults to False. Then tests will tell us when we need to set it to True.